internal/grpctest/tlogger.go

```diff
@@ -63,25 +65,33 @@ func init() {
 	}
 }
 
-// getStackFrame gets, from the stack byte string, the appropriate stack frame.
-func getStackFrame(stack []byte, frame int) (string, error) {
-	s := strings.Split(string(stack), "\n")
-	if frame >= (len(s)-1)/2 {
+// getCallingPrefix returns the <file:line> at the given depth from the stack.
+func getCallingPrefix(depth int) (string, error) {
+	_, file, line, ok := runtime.Caller(depth)
+	if !ok {
 		return "", errors.New("frame request out-of-bounds")
 	}
-	split := strings.Split(strings.Fields(s[(frame*2)+2][1:])[0], "/")
-	return fmt.Sprintf("%v:", split[len(split)-1]), nil
+	return fmt.Sprintf("%s:%d", path.Base(file), line), nil
+}
+
+// Returns the last component of the stringified current time, which is of the
+// format "m=±<value>", where value is the monotonic clock reading formatted as
+// a decimal number of seconds.
+func getTimeSuffix() string {
+	parts := strings.Split(time.Now().String(), " ")
```

Comment:
Can the time not have a monotonic clock reading? Is it always there in the newer versions of go?

How is this going to be used? Can you give an example?

And does tlogger already print the time? If no, why not include the full string?

Reply:
The test output would look like this:
```
=== RUN   Test/V2ClientBackoffAfterRecvError
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:249 [core] parsed scheme: ""  (m=+18.201969668)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:255 [core] scheme "" not registered, fallback to default scheme  (m=+18.201988944)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: resolver_conn_wrapper.go:143 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:61559  <nil> 0 <nil>}] <nil> <nil>}  (m=+18.202009503)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:681 [core] ClientConn switching balancer to "pick_first"  (m=+18.202022420)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:696 [core] Channel switches to new LB policy "pick_first"  (m=+18.202034162)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:1073 [core] Subchannel Connectivity change to CONNECTING  (m=+18.202054932)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:1210 [core] Subchannel picks a new address "127.0.0.1:61559" to connect  (m=+18.202076470)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:428 [core] Channel Connectivity change to CONNECTING  (m=+18.202087040)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:1073 [core] Subchannel Connectivity change to READY  (m=+18.202498720)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:428 [core] Channel Connectivity change to READY  (m=+18.202528586)
    Test/V2ClientBackoffAfterRecvError: client_test.go:536: Started xds v2Client...
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: transport_helper.go:205 ADS stream created  (m=+18.202648378)
    Test/V2ClientBackoffAfterRecvError: client_test.go:544: FakeServer received request...
    Test/V2ClientBackoffAfterRecvError: client_test.go:547: Bad LDS response pushed to fakeServer...
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: transport_helper.go:311 ADS stream is closed with error: xds: stream.Recv() failed: rpc error: code = Unknown desc = RPC error  (m=+18.202925006)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: transport_helper.go:205 ADS stream created  (m=+18.202960780)
    Test/V2ClientBackoffAfterRecvError: client_test.go:555: v2Client backed off before retrying...
    Test/V2ClientBackoffAfterRecvError: client_test.go:563: FakeServer received request after backoff...
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:428 [core] Channel Connectivity change to SHUTDOWN  (m=+18.203163928)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:1073 [core] Subchannel Connectivity change to CONNECTING  (m=+18.203179736)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: transport_helper.go:311 ADS stream is closed with error: xds: stream.Recv() failed: rpc error: code = Unavailable desc = transport is closing  (m=+18.203251344)
    Test/V2ClientBackoffAfterRecvError: tlogger.go:107: clientconn.go:1073 [core] Subchannel Connectivity change to SHUTDOWN  (m=+18.203244811)
```

Although, I just realized that this doesn't start from zero for every subtest. I think it just reset only per test binary (which might be one per package). This is an alternative to having the full timestamp since you can see how much time has elapsed from the last print. But I'm OK to just print the whole timestamp as well (it will make it a little simpler since we wont have to worry about systems which dont have a monotonic clock).

Reply:
>How is this going to be used? Can you give an example?

There are times when you look at the test output, and something has timed out. And without any timestamps in the output, there is no way for you to figure out which step consumed how much time.

Reply:
Overall SGTM.

Two other thoughts:
1. Move to the beginning of the line? (Not sure which I like more...)
2. Keep `time.Now()` at the beginning of the test, and log `time.Since()`?

Reply:
> 1. Move to the beginning of the line? (Not sure which I like more...)
I think it might be better to be at the end of the line, since we would look at the preceding text more often than the timestamp.

> 2. Keep time.Now() at the beginning of the test, and log time.Since()?
Will do that in a follow-up PR. I just noticed that `TLogger` being a global var is not very appropriate here. I will change that to be variable inside the `grpctest.Tester` struct, and then store the test start time there, and do `time.Since()`.

Thanks.